### PR TITLE
feat: Add JsonSchema package.

### PR DESCRIPTION
**Pull request by Romakita**
## Informations

Type | Status | Migration
---|---|---
Feature | ready | No

****

## Description
All class decorated with JsonProperty generate a JsonSchema (base on http://json-schema.org/). This JsonSchema can be used by a validator package and swagger.

## Usage example
```typescript

export class JsonFoo2 {
    @JsonProperty()
    @Required()
    test: string;

    @JsonProperty("Name")
    name: string;

    @JsonProperty()
    dateStart: Date;

    @JsonProperty()
    uint: number;

    @JsonProperty()
    object: any;

    @JsonProperty()
    @Required()
    foo: JsonFoo;

    @JsonProperty({use: JsonFoo})
    foos: JsonFoo[];

    @JsonProperty({use: JsonFoo})
    theMap: Map<string, JsonFoo>;

    @JsonProperty({use: JsonFoo})
    theSet: Set<JsonFoo>;

    @JsonProperty({use: String})
    mapOfString: Map<string, string>;

    @JsonProperty({use: String})
    arrayOfString: string[];

    @JsonProperty()
    nameModel: JsonNameModel;

    @JsonProperty()
    ageModel: JsonAgeModel;

    method() {

    }
}
```
Will generate this JsonSchema:
```json
{
            "definitions": {
                "JsonAgeModel": {
                    "properties": {
                        "age": {
                            "type": "number"
                        }
                    },
                    "type": "object"
                },
                "JsonNameModel": {
                    "properties": {
                        "name": {
                            "type": "string"
                        }
                    },
                    "type": "object"
                }
            },
            "properties": {
                "ageModel": {
                    "$ref": "#/definitions/JsonAgeModel"
                },
                "arrayOfString": {
                    "items": {
                        "type": "string"
                    },
                    "type": "array"
                },
                "dateStart": {
                    "type": "string"
                },
                "foo": {
                    "type": "object"
                },
                "foos": {
                    "items": {
                        "type": "object"
                    },
                    "type": "array"
                },
                "mapOfString": {
                    "additionalProperties": {
                        "type": "string"
                    }
                },
                "name": {
                    "type": "string"
                },
                "nameModel": {
                    "$ref": "#/definitions/JsonNameModel"
                },
                "object": {
                    "type": "object"
                },
                "test": {
                    "type": "string"
                },
                "theMap": {
                    "additionalProperties": {
                        "type": "object"
                    }
                },
                "theSet": {
                    "additionalProperties": {
                        "type": "object"
                    }
                },
                "uint": {
                    "type": "number"
                }
            },
            "required": [
                "test",
                "foo"
            ],
            "type": "object"
        }
```
## Todos

- [x] Tests
- [x] Coverage
- [x] Example
~~- [ ] Documentation~~

### CODE REVIEW

File: src/jsonschema/class/PropertyMetadata.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+    /**
+     * Set the allowed values when the value is required.
+     * @param {any[]} value
+     */
+    set allowedValues(value: any[]) {
+        this._allowedValues = value;
+    }
+
+    isValidValue(value: any): boolean {
```
Comment (lwallent) [inline src/jsonschema/class/PropertyMetadata.ts]: Should this method be renamed to `isValidRequiredValue` as the validity only has to do with it being required? Same goes for the `allowedValues` witch could be named `allowedRequiredValues`. Not really sure about the usage - so just a thought for consideration :-)

File: test/units/jsonschema/class/PropertyMetadata.spec.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    });
+
+    describe("isValidValue", () => {
+        describe("when property is required", () => {
+            before(() => {
+                this.propertyMetadata = new PropertyMetadata(Test, "test");
+                this.propertyMetadata.allowedValues = [];
+                this.propertyMetadata.required = true;
+            });
+            it("should return false (value 0)", () => {
```
Comment (lwallent) [inline test/units/jsonschema/class/PropertyMetadata.spec.ts]: Are all these tests not upside-down? 
"should return false .."  - expect(THE RETURN VALUE).to.be.true ??
"should return true .."   - expect(THE RETURN VALUE).to.be.false ??

Or perhaps it is just my head spinning :-)

Comment (Romakita) [reply test/units/jsonschema/class/PropertyMetadata.spec.ts]: arf... bad copy paste XD
I'll check that. thanks ;)

File: src/jsonschema/class/PropertyMetadata.ts (hunk 1 above)
Comment (Romakita) [reply src/jsonschema/class/PropertyMetadata.ts]: I agree ;)

**Comment by Romakita** [comment]
done :)
Thanks @lwallent ;)

**Pull request merged**